src/qunit.cpp

```diff
@@ -38,19 +38,42 @@ QUnit::QUnit(QInterfaceEngine eng, bitLenInt qBitCount, bitCapInt initState, std
     }
 }
 
+void QUnit::CopyState(QUnitPtr orig)
+{
+    shards.clear();
+    SetQubitCount(orig->GetQubitCount());
+
+    /* Set up the shards to refer to the new unit. */
+    std::map<QInterfacePtr, QInterfacePtr> otherUnits;
+    for (auto otherShard :orig->shards) {
```

Comment:
This simplifies a bit if you use a classical for loop:
```
    for (int i = 0; i < orig->numQBits; i++) {
        auto engine = otherUnits.find(orig->shards[i].unit);
        if (engine == otherUnits.end()) {
            otherUnits[orig->shards[i].unit] = shards[i].unit = CreateQuantumInterface(...);
        } else {
            shards[i].unit = engine.second;
        }
        shards[i].mapped = orig->shards[i].mapped;
    }
```

Reply:
You still need the push_back, actually, because the method starts by clearing the vector. If the vector is cleared and pushed back, the method is safe to use on QInterfacePtr instances that point to units of different sizes. Adding back in the declaration and the push_back, it's really just about as complicated, and I don't think it makes any significant difference to speed, looking at both versions. We could reserve space in the vector, but that's kind of overkill, and it's probably still allocated to appropriate size after the clear().